src/lua/functions/core/libs/kv_functions.cpp

```diff
@@ -0,0 +1,138 @@
+	// KV.set(key, value) | scopedKV:set(key, value)
+	auto key = getString(L, -2);
+	auto valueWrapper = getValueWrapper(L);
+
+	if (!valueWrapper) {
+		g_logger().warn("[{}] invalid param type", __FUNCTION__);
+		pushBoolean(L, false);
+		return 1;
+	}
+
+	if (isUserdata(L, 1)) {
+		auto scopedKV = getUserdata<KVStore::ScopedKV>(L, 1);
+		scopedKV->set(key, valueWrapper.value());
+		pushBoolean(L, true);
+		return 1;
+	}
+
+	g_kv().set(key, valueWrapper.value());
+	pushBoolean(L, true);
+	return 1;
+}
+
+int KVFunctions::luaKVGet(lua_State* L) {
+	// KV.get(key) | scopedKV:get(key)
+	std::optional<ValueWrapper> valueWrapper;
+	auto key = getString(L, -1);
+	if (isUserdata(L, 1)) {
+		auto scopedKV = getUserdata<KVStore::ScopedKV>(L, 1);
+		valueWrapper = scopedKV->get(key);
+	} else {
+		valueWrapper = g_kv().get(key);
+	}
+
+	if (valueWrapper) {
+		pushValueWrapper(L, *valueWrapper);
+	} else {
+		lua_pushnil(L);
+	}
+	return 1;
+}
+
+std::optional<ValueWrapper> KVFunctions::getValueWrapper(lua_State* L) {
+	if (isString(L, -1)) {
+		return getString(L, -1);
+	} else if (isNumber(L, -1)) {
+		return getNumber<double>(L, -1);
+	} else if (isBoolean(L, -1)) {
+		return getBoolean(L, -1);
+	} else if (isTable(L, -1)) {
+		if (lua_objlen(L, -1) > 0) {
+			ArrayType array;
+			for (int i = 1; i <= lua_objlen(L, -1); ++i) {
+				lua_rawgeti(L, -1, i);
+				auto value = getValueWrapper(L);
+				if (!value) {
+					g_logger().warn("[{}] invalid param type", __FUNCTION__);
+					return std::nullopt;
+				}
+				array.push_back(value.value());
+				lua_pop(L, 1);
+			}
+			return ValueWrapper(array);
+		} else {
+			MapType map;
+			lua_pushnil(L);
+			while (lua_next(L, -2) != 0) {
+				auto value = getValueWrapper(L);
+				if (!value) {
+					g_logger().warn("[{}] invalid param type", __FUNCTION__);
+					return std::nullopt;
+				}
+				map[getString(L, -2)] = std::make_shared<ValueWrapper>(value.value());
+				lua_pop(L, 1);
+			}
+			return ValueWrapper(map);
+		}
+	} else if (isNil(L, -1)) {
+		return nullptr;
+	} else {
+		g_logger().warn("[{}] invalid param type", __FUNCTION__);
+		return std::nullopt;
+	}
+}
+
+void KVFunctions::pushValueWrapper(lua_State* L, ValueWrapper &valueWrapper) {
+	std::visit(
+		[&](auto &&arg) {
+			using T = std::decay_t<decltype(arg)>;
+			if constexpr (std::is_same_v<T, StringType>) {
+				pushString(L, arg);
+			} else if constexpr (std::is_same_v<T, IntType>) {
+				lua_pushnumber(L, arg);
+			} else if constexpr (std::is_same_v<T, DoubleType>) {
+				lua_pushnumber(L, arg);
+			} else if constexpr (std::is_same_v<T, ArrayType>) {
+				lua_newtable(L);
+				for (int i = 0; i < arg.size(); ++i) {
+					pushValueWrapper(L, arg[i]);
+					lua_rawseti(L, -2, i + 1);
+				}
+			} else if constexpr (std::is_same_v<T, MapType>) {
+				lua_newtable(L);
+				for (const auto &[key, value] : arg) {
+					pushValueWrapper(L, *value);
+					lua_setfield(L, -2, key.c_str());
+				}
+			}
+		},
+		valueWrapper.getVariant()
+	);
```

Comment:
Same here, you could add return; in each if and remove else branches to improve reading.